File: sdk/typescript/src/namespaces/sso.ts
/**
 * SSO (Single Sign-On) Namespace API
 *
 * Provides endpoints for enterprise SSO authentication via SAML 2.0 and OIDC.
 */

import type { AragoraClient } from '../client';

/**
 * SSO provider types
 */
export type SSOProviderType = 'saml' | 'oidc' | 'oauth2';

/**
 * SSO login response with redirect URL
 */
export interface SSOLoginResponse {
  redirect_url: string;
  state?: string;
  nonce?: string;
  provider: SSOProviderType;
}

/**
 * SSO user information from identity provider
 */
export interface SSOUser {
  id: string;
  email: string;
  name?: string;
  given_name?: string;
  family_name?: string;
  groups?: string[];
  roles?: string[];
  organization_id?: string;
  attributes?: Record<string, unknown>;
  provider: SSOProviderType;
  provider_id?: string;
}

/**
 * SSO callback result after IdP authentication
 */
export interface SSOCallbackResult {
  success: boolean;
  user?: SSOUser;
  token?: string;
  refresh_token?: string;
  expires_in?: number;
  error?: string;
  error_description?: string;
}

/**
 * SSO provider status
 */
export interface SSOStatus {
  enabled: boolean;
  provider_type?: SSOProviderType;
  idp_url?: string;
  entity_id?: string;
  configured: boolean;
  last_login?: string;
  total_logins?: number;
}

/**
 * SAML SP metadata
 */
export interface SAMLMetadata {
  entity_id: string;
  acs_url: string;
  slo_url?: string;
  certificate?: string;
  metadata_xml?: string;
}

/**
 * SSO logout response
 */
export interface SSOLogoutResponse {
  success: boolean;
  redirect_url?: string;
  message?: string;
}

/**
 * SSO namespace for Single Sign-On authentication.
 *
 * @example
 * ```typescript
 * // Check if SSO is configured
 * const status = await client.sso.getStatus();
 * if (status.enabled) {
 *   // Initiate SSO login
 *   const login = await client.sso.login({ returnUrl: '/dashboard' });
 *   window.location.href = login.redirect_url;
 * }
 *
 * // Get SAML metadata for IdP configuration
 * const metadata = await client.sso.getMetadata();
 * ```
 */
export class SSONamespace {
  constructor(private client: AragoraClient) {}

  /**
   * Get SSO configuration status.
   *
   * Returns whether SSO is enabled and configured.
   */
  async getStatus(): Promise<SSOStatus> {
    return this.client.request<SSOStatus>('GET', '/auth/sso/status');
  }

  /**
   * Initiate SSO login flow.
   *
   * Returns a redirect URL to the identity provider.
   *
   * @param options.returnUrl - URL to redirect after successful login
   * @param options.prompt - OIDC prompt parameter (e.g., 'login', 'consent')
   */
  async login(options?: { returnUrl?: string; prompt?: string }): Promise<SSOLoginResponse> {
    const params = new URLSearchParams();
    if (options?.returnUrl) params.set('return_url', options.returnUrl);
    if (options?.prompt) params.set('prompt', options.prompt);

    const query = params.toString();
    const path = query ? `/auth/sso/login?${query}` : '/auth/sso/login';
    return this.client.request<SSOLoginResponse>('GET', path);
  }

  /**
   * Handle SSO callback from identity provider.
   *
   * This is typically called by the server, not the client directly.
   * Use this to validate and exchange the authorization code.
   *
   * @param params - Callback parameters from IdP (code, state, etc.)
   */
  async callback(params: {
    code?: string;
    state?: string;
    SAMLResponse?: string;
    RelayState?: string;
  }): Promise<SSOCallbackResult> {
    return this.client.request<SSOCallbackResult>('POST', '/auth/sso/callback', { body: params });
  }

  /**
   * Initiate SSO logout.
   *
   * @param options.everywhere - Log out from all sessions (SLO)
   */
  async logout(options?: { everywhere?: boolean }): Promise<SSOLogoutResponse> {
    const body = options?.everywhere ? { everywhere: true } : undefined;
    return this.client.request<SSOLogoutResponse>('POST', '/auth/sso/logout', { body });
  }

  /**
   * Get SAML Service Provider metadata.
   *
   * Returns XML metadata for configuring your Identity Provider.
   * Only available for SAML-based SSO.
   */
  async getMetadata(): Promise<SAMLMetadata> {
    return this.client.request<SAMLMetadata>('GET', '/auth/sso/metadata');
  }

  /**
   * Get raw SAML metadata XML.
   *
   * @returns SAML SP metadata as XML string
   */
  async getMetadataXML(): Promise<string> {
    const metadata = await this.getMetadata();
    return metadata.metadata_xml || '';
  }

  /**
   * Check if SSO is enabled.
   */
  async isEnabled(): Promise<boolean> {
    const status = await this.getStatus();
    return status.enabled;
  }

  /**
   * Check if SSO is properly configured.
   */
  async isConfigured(): Promise<boolean> {
    const status = await this.getStatus();
    return status.configured;
  }
}

File: sdk/python/aragora_sdk/namespaces/sso.py
"""
SSO (Single Sign-On) Namespace API

Provides methods for enterprise SSO authentication:
- Check SSO configuration status
- Initiate SSO login flow
- Handle SSO callback from identity provider
- Logout from SSO session
- Retrieve SAML/OIDC provider metadata
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class SSOAPI:
    """
    Synchronous SSO API.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> status = client.sso.get_status()
        >>> if status.get("enabled"):
        ...     login = client.sso.login()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_status(self) -> dict[str, Any]:
        """Get SSO configuration status."""
        return self._client.request("GET", "/api/v2/sso/status")

    def login(self) -> dict[str, Any]:
        """Initiate SSO login flow."""
        return self._client.request("GET", "/api/v2/sso/login")

    def callback(self) -> dict[str, Any]:
        """Handle SSO callback (GET)."""
        return self._client.request("GET", "/api/v2/sso/callback")

    def callback_post(self, **kwargs: Any) -> dict[str, Any]:
        """Handle SSO callback (POST)."""
        return self._client.request("POST", "/api/v2/sso/callback", json=kwargs)

    def logout(self) -> dict[str, Any]:
        """Logout from SSO session."""
        return self._client.request("GET", "/api/v2/sso/logout")

    def get_metadata(self) -> dict[str, Any]:
        """Get SSO provider metadata."""
        return self._client.request("GET", "/api/v2/sso/metadata")

    def sso_login(self) -> dict[str, Any]:
        """Initiate SSO login via auth endpoint.

        Returns:
            Dict with SSO login redirect URL and session data.
        """
        return self._client.request("GET", "/auth/sso/login")

    def auth_callback(self) -> dict[str, Any]:
        """Handle SSO callback via auth endpoint (GET)."""
        return self._client.request("GET", "/auth/sso/callback")

    def auth_callback_post(self, **kwargs: Any) -> dict[str, Any]:
        """Handle SSO callback via auth endpoint (POST)."""
        return self._client.request("POST", "/auth/sso/callback", json=kwargs)

    def auth_logout(self) -> dict[str, Any]:
        """Logout from SSO via auth endpoint."""
        return self._client.request("GET", "/auth/sso/logout")

    def auth_metadata(self) -> dict[str, Any]:
        """Get SSO provider metadata via auth endpoint."""
        return self._client.request("GET", "/auth/sso/metadata")

    def auth_status(self) -> dict[str, Any]:
        """Get SSO configuration status via auth endpoint."""
        return self._client.request("GET", "/auth/sso/status")


class AsyncSSOAPI:
    """
    Asynchronous SSO API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     status = await client.sso.get_status()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_status(self) -> dict[str, Any]:
        """Get SSO configuration status."""
        return await self._client.request("GET", "/api/v2/sso/status")

    async def login(self) -> dict[str, Any]:
        """Initiate SSO login flow."""
        return await self._client.request("GET", "/api/v2/sso/login")

    async def callback(self) -> dict[str, Any]:
        """Handle SSO callback (GET)."""
        return await self._client.request("GET", "/api/v2/sso/callback")

    async def callback_post(self, **kwargs: Any) -> dict[str, Any]:
        """Handle SSO callback (POST)."""
        return await self._client.request("POST", "/api/v2/sso/callback", json=kwargs)

    async def logout(self) -> dict[str, Any]:
        """Logout from SSO session."""
        return await self._client.request("GET", "/api/v2/sso/logout")

    async def get_metadata(self) -> dict[str, Any]:
        """Get SSO provider metadata."""
        return await self._client.request("GET", "/api/v2/sso/metadata")

    async def sso_login(self) -> dict[str, Any]:
        """Initiate SSO login via auth endpoint."""
        return await self._client.request("GET", "/auth/sso/login")

    async def auth_callback(self) -> dict[str, Any]:
        """Handle SSO callback via auth endpoint (GET)."""
        return await self._client.request("GET", "/auth/sso/callback")

    async def auth_callback_post(self, **kwargs: Any) -> dict[str, Any]:
        """Handle SSO callback via auth endpoint (POST)."""
        return await self._client.request("POST", "/auth/sso/callback", json=kwargs)

    async def auth_logout(self) -> dict[str, Any]:
        """Logout from SSO via auth endpoint."""
        return await self._client.request("GET", "/auth/sso/logout")

    async def auth_metadata(self) -> dict[str, Any]:
        """Get SSO provider metadata via auth endpoint."""
        return await self._client.request("GET", "/auth/sso/metadata")

    async def auth_status(self) -> dict[str, Any]:
        """Get SSO configuration status via auth endpoint."""
        return await self._client.request("GET", "/auth/sso/status")
